platform/consul/controller_test.go

```diff
@@ -0,0 +1,219 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package consul
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/hashicorp/consul/api"
+	"istio.io/pilot/model"
+)
+
+var (
+	services = map[string][]string{
+		"productpage": {"version=v1"},
+		"reviews":     {"version=v1", "version=v2", "version=v3"},
+	}
+
+	productpage = []*api.CatalogService{
+		{
+			Node:           "istio",
+			Address:        "172.19.0.5",
+			ServiceID:      "111-111-111",
+			ServiceName:    "productpage",
+			ServiceTags:    []string{"version=v1"},
+			ServiceAddress: "172.19.0.11",
+			ServicePort:    9080,
+		},
+	}
+	reviews = []*api.CatalogService{
+		{
+			Node:           "istio",
+			Address:        "172.19.0.5",
+			ServiceID:      "222-222-222",
+			ServiceName:    "reviews",
+			ServiceTags:    []string{"version=v1"},
+			ServiceAddress: "172.19.0.6",
+			ServicePort:    9080,
+		},
+		{
+			Node:           "istio",
+			Address:        "172.19.0.5",
+			ServiceID:      "333-333-333",
+			ServiceName:    "reviews",
+			ServiceTags:    []string{"version=v2"},
+			ServiceAddress: "172.19.0.7",
+			ServicePort:    9080,
+		},
+		{
+			Node:           "istio",
+			Address:        "172.19.0.5",
+			ServiceID:      "444-444-444",
+			ServiceName:    "reviews",
+			ServiceTags:    []string{"version=v3"},
+			ServiceAddress: "172.19.0.8",
+			ServicePort:    9080,
+			NodeMeta:       map[string]string{protocolTagName: "tcp"},
+		},
+	}
+)
+
+func newServer() *httptest.Server {
```

Comment:
could reduce duplicate code by passing in `t *testing.T` and `t.Fail(err)` if `err != nil` inside `newServer`